fix(mobile): handle readonly and offline assets

shenlong-tanwen opened the pull request:
#### Changes made in the PR

- Refactored the filters on `Iterable<Asset>`  to an extension to make it accessible across the app
- Store the fields `isOffline` and `isReadOnly` for remote asssets
- Updated the actions across the app to disable the following based on the above two fields:
    -  Delete & Edit meta is not available for `isReadOnly` assets. This is because all readOnly assets are assumed to be in ro mounts and we'd have no way to write / modify files
    - Share & Download is not available for `isOffline` assets since we cannot access the underlying file

Fixes #6204

shenlong-tanwen committed:
feat: add isReadOnly and isOffline fields to Asset collection

shenlong-tanwen committed:
refactor: move asset iterable filters to extension

shenlong-tanwen committed:
hide asset actions based on offline and readOnly fields

fyfrey commented on mobile/lib/modules/asset_viewer/views/gallery_viewer.dart:
```diff
@@ -332,6 +351,26 @@ class GalleryViewerPage extends HookConsumerWidget {
+        return;
+      }
+
+      ref.watch(imageViewerStateProvider.notifier).downloadAsset(
```
Sorry for the spam: use read instead of watch in callbacks?

shenlong-tanwen replied:
Ah, not a problem 😄 Will update all such instances soon. Thank you!

fyfrey commented on mobile/lib/extensions/collection_extensions.dart:
```diff
@@ -39,3 +41,52 @@ extension IntListExtension on Iterable<int> {
   }
 }
+
+extension AssetListExtension on Iterable<Asset> {
```
Great idea!

fyfrey reviewed (commented):
Good fix for those assets. I think this needs a DB migration (just truncate the tables like the other migrations) to add the new fields to asset with the proper values from the server. Otherwise, users need to perform a manual refresh

shenlong-tanwen committed:
pr changes

shenlong-tanwen commented:
> Good fix for those assets. I think this needs a DB migration (just truncate the tables like the other migrations) to add the new fields to asset with the proper values from the server. Otherwise, users need to perform a manual refresh

Ah, I'll keep in mind to migrate Isar moving forwards on new fields. I've updated the change-set with migration and also updated all the notifier watch calls to read :)

alextran1502 commented on mobile/lib/extensions/collection_extensions.dart:
```diff
@@ -39,3 +41,52 @@ extension IntListExtension on Iterable<int> {
+    return this;
+  }
+
+  Iterable<Asset> liveOnly({
```
I assume live here is LivePhotos?

shenlong-tanwen replied:
It was actually to filter out non offline files, so updated the method name to `nonOfflineOnly` to better indicate the same

alextran1502 commented on mobile/lib/extensions/collection_extensions.dart:
```diff
@@ -39,3 +41,52 @@ extension IntListExtension on Iterable<int> {
+    return this;
+  }
+
+  Iterable<Asset> writableOnly({
```
What does indicate that an asset is writeable only? Can we add some comments on these methods for documentation purposes?

shenlong-tanwen replied:
Done :)

shenlong-tanwen committed:
chore: doc comments

alextran1502 committed:
merge main
